logging/src/lib.rs

```diff
@@ -0,0 +1,340 @@
+    /*  7 */ GuardedMessage(GuardedMessageEvent),
+    /*  8 */ GuardedProgress(GuardedProgressEvent),
+    /*  9 */ CommChannels(CommChannelsEvent),
+    /* 10 */ Input(InputEvent),
+}
+
+#[derive(Debug, Clone, Abomonation)]
+pub enum CommsEvent {
+    /*  0 */ Communication(CommunicationEvent),
+    /*  1 */ Serialization(SerializationEvent),
+}
+
+impl From<CommunicationEvent> for CommsEvent {
+    fn from(v: CommunicationEvent) -> CommsEvent { CommsEvent::Communication(v) }
+}
+
+impl From<SerializationEvent> for CommsEvent {
+    fn from(v: SerializationEvent) -> CommsEvent { CommsEvent::Serialization(v) }
+}
+
+impl From<OperatesEvent> for Event {
+    fn from(v: OperatesEvent) -> Event { Event::Operates(v) }
+}
+
+impl From<ChannelsEvent> for Event {
+    fn from(v: ChannelsEvent) -> Event { Event::Channels(v) }
+}
+
+impl From<ProgressEvent> for Event {
+    fn from(v: ProgressEvent) -> Event { Event::Progress(v) }
+}
+
+impl From<PushProgressEvent> for Event {
+    fn from(v: PushProgressEvent) -> Event { Event::PushProgress(v) }
+}
+
+impl From<MessagesEvent> for Event {
+    fn from(v: MessagesEvent) -> Event { Event::Messages(v) }
+}
+
+impl From<ScheduleEvent> for Event {
+    fn from(v: ScheduleEvent) -> Event { Event::Schedule(v) }
+}
+
+impl From<ApplicationEvent> for Event {
+    fn from(v: ApplicationEvent) -> Event { Event::Application(v) }
+}
+
+impl From<GuardedMessageEvent> for Event {
+    fn from(v: GuardedMessageEvent) -> Event { Event::GuardedMessage(v) }
+}
+
+impl From<GuardedProgressEvent> for Event {
+    fn from(v: GuardedProgressEvent) -> Event { Event::GuardedProgress(v) }
+}
+
+impl From<CommChannelsEvent> for Event {
+    fn from(v: CommChannelsEvent) -> Event { Event::CommChannels(v) }
+}
+
+impl From<InputEvent> for Event {
+    fn from(v: InputEvent) -> Event { Event::Input(v) }
+}
+
+const BUFFERING_LOGGER_CAPACITY: usize = 1024;
+
+pub enum LoggerBatch<'a, S: Clone+'a, L: Clone+'a> {
+    Logs(&'a Vec<(u64, S, L)>),
+    End,
+}
+
+enum BufferingLoggerInternal<S: Clone, L: Clone> {
+    Active {
+        setup: S,
+        buffer: RefCell<Vec<(u64, S, L)>>,
+        pushers: RefCell<Box<Fn(LoggerBatch<S, L>)->()>>,
+    },
+    Inactive
+}
+
+pub struct BufferingLogger<S: Clone, L: Clone> {
+    internal: BufferingLoggerInternal<S, L>,
+}
+
+impl<S: Clone, L: Clone> BufferingLogger<S, L> {
+    pub fn new(setup: S, pushers: Box<Fn(LoggerBatch<S, L>)->()>) -> Self {
+        BufferingLogger {
+            internal: BufferingLoggerInternal::Active {
+                setup: setup,
+                buffer: RefCell::new(Vec::with_capacity(BUFFERING_LOGGER_CAPACITY)),
+                pushers: RefCell::new(pushers),
+            },
+        }
+    }
+
+    pub fn new_inactive() -> Rc<BufferingLogger<EventsSetup, Event>> {
+        Rc::new(BufferingLogger {
+            internal: BufferingLoggerInternal::Inactive,
+        })
+    }
+
+    pub fn log(&self, l: L) {
+        match self.internal {
```

Comment:
Is a particular reason to have the `RefCell` inside the `BufferingLogger` and to use `&self`, versus having the `Logger` type be `Rc<RefCell<BufferingLogger>>` and have this take a `&mut self`? Not judging, just trying to grok whether we are intentionally hiding the potential sharing and creating a risk (e.g. a logger that logs could panic, right?), or if it would work as well with `&mut` and the `RefCell` lifted up a level.